### PR TITLE
[KYUUBI #4576] Multi arch build for ARM support in Docker Hub

### LINKED ISSUE #4576: [Improvement] Multi Arch build for ARM Support in Docker Hub
### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### Search before asking

- [X] I have searched in the [issues](https://github.com/apache/kyuubi/issues?q=is%3Aissue) and found no similar issues.

### What would you like to be improved?

Sorry, I'm late for busy.

Based on This [Discussion](https://github.com/apache/kyuubi/discussions/4487)

> Hi, I would like to request a new feature that provides support for ARM-based images Apache Kyuubi.
Currently, in the [Docker Hub](https://hub.docker.com/r/apache/kyuubi/tags) only provides images for x86-based architectures.

> Have you any plan of ARM architecture Image support?

ARM support at docker hub

### How should we improve?

Update [.github/workflows/publish-snapshot-docker.yml](https://github.com/apache/kyuubi/blob/824aba60708c0973ee638eeb97d296e2f8dc737e/.github/workflows/publish-snapshot-docker.yml)

I'm already have commit that implements.
https://github.com/phyyou/kyuubi/tree/feat/buildx-at-docker-hub-snapshot

https://github.com/apache/kyuubi/pull/4577

However, I think need change https://github.com/apache/kyuubi-docker this repo also for fully improve and support arm image at docker hub

### Are you willing to submit PR?

- [X] Yes. I would be willing to submit a PR with guidance from the Kyuubi community to improve.
- [ ] No. I cannot submit a PR at this time.

### PR DESCRIPTION
**Pull request by phyyou**
Close #4576 

* Setup QEMU and Buildx.
* Use action docker/build-push-action@v4.
* build multi platform (amd64, arm64)
* build cache for faster build ✨

### _Why are the changes needed?_

Based on This [Disscussion](https://github.com/apache/kyuubi/discussions/4487) and [Issue](https://github.com/apache/kyuubi/issues/4576)

Update publish-snapshot-docker workflow

### _How was this patch tested?_
now test in progress...

image

https://github.com/phyyou/_kyuubi/actions/runs/4485366649/jobs/7886838765

- [x] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.readthedocs.io/en/master/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

**Review by pan3793 — APPROVED** [review]

**Comment by pan3793** [comment]
Is something wrong? Looks like the job took 4 hours but not completed

File: .github/workflows/publish-snapshot-docker.yml [hunk 1]
```diff
@@ -29,14 +29,22 @@ jobs:
...
+      - name: Build and Push Kyuubi Docker Image
+        uses: docker/build-push-action@v4
+        with:
+          context: .
+          platforms: linux/amd64,linux/arm64
+          push: true
+          file: build/Dockerfile
+          cache-from: type=gha
+          cache-to: type=gha,mode=max
+          tags: apache/kyuubi:master-snapshot
```
Comment (bowenliang123) [inline .github/workflows/publish-snapshot-docker.yml:50]: Reorder configs in alphabetical order or in descending order of importance.

Comment (bowenliang123) [inline .github/workflows/publish-snapshot-docker.yml:50]: Reorder configs in alphabetical order or in descending order of importance.

File: .github/workflows/publish-snapshot-docker.yml [hunk 2]
```diff
@@ -29,14 +29,22 @@ jobs:
...
-      - name: Push Docker image
-        run: docker push apache/kyuubi:master-snapshot
+      - name: Build and Push Kyuubi Docker Image
+        uses: docker/build-push-action@v4
+        with:
+          context: .
+          platforms: linux/amd64,linux/arm64
+          push: true
+          file: build/Dockerfile
+          cache-from: type=gha
```
Comment (bowenliang123) [inline .github/workflows/publish-snapshot-docker.yml:48]: What does `gha` stands for? Didn't find a clue on https://github.com/docker/build-push-action .

**Review by bowenliang123 — APPROVED** [review]
LGTM overall. Minor suggestion and question.

File: .github/workflows/publish-snapshot-docker.yml (hunk 2 above)
Comment (phyyou) [reply .github/workflows/publish-snapshot-docker.yml:48]: It is cache on Github Actions
Docs: https://docs.docker.com/build/cache/backends/gha/#using-dockerbuild-push-action

Comment (bowenliang123) [reply .github/workflows/publish-snapshot-docker.yml:48]: Thanks for the hint. Better to attach this information to the description for future reference.

**Comment by phyyou** [comment]
> Is something wrong? Looks like the job took 4 hours but not completed

@pan3793 Yup. It just completed with 4 hours. it seems maven build took long times in buildx and qemu.

So I add build cache at Github Action for faster.

File: .github/workflows/publish-snapshot-docker.yml (hunk 1 above)
Comment (phyyou) [reply .github/workflows/publish-snapshot-docker.yml:50]: Ok

File: .github/workflows/publish-snapshot-docker.yml (hunk 2 above)
Comment (phyyou) [reply .github/workflows/publish-snapshot-docker.yml:48]: Got it. thanks for suggestion.

**Comment by pan3793** [comment]
@phyyou how long will take in case of a cache hit? but anyway, it's a daily job, 4 hours is acceptable

**Comment by phyyou** [comment]
> @phyyou how long will take in case of a cache hit? but anyway, it's a daily job, 4 hours is acceptable

@pan3793 
https://github.com/phyyou/_kyuubi/actions/runs/4485366649/jobs/7891661788

I'm not sure when update deps of maven build. but it has all cached layer. It tooks 11 seconds.

**Comment by pan3793** [comment]
Thanks, merged to master

**Comment by ulysses-you** [comment]
thank you @phyyou , are you interested in publish release https://github.com/apache/kyuubi-docker ? We can make the stable tag support ARM.

**Comment by phyyou** [comment]
Thanks @ulysses-you.
Yes, I would be happy and want to use stable release image in Docker Hub that supports ARM architecture.
I'll stay tuned for updates.

**Pull request closed without merging**